learningequality/kolibri
# List Users Login Flow in App Context

### Summary


Revises the Sign In flow per designs linked below in References.

- When in app context and <= 16 users in Facility, we list the users by name which simulates typing a username.
- If a password is NOT required - clicking a user in a list (or entering your username) will log you in.
- If a password IS REQUIRED - clicking a user in a list (or entering your username) will present you with a password text field, which will log you in or show errors.
- If a user does not have a password - but the setting is changed to requiring a password, the user is asked to create a password. The user is logged in - their password is updated - then they're redirected to wherever they were going to in the first place.

### Reviewer guidance


*Setup*

- Have a device with <= 16 total users in a facility and select that facility at the sign-in page.
- One or more of those users must be a Learner. 
- Have your handy [SQLite browser](https://sqlitebrowser.org/) app and open your database in it. In this browser there should be a "Browse Data" tab. Then a drop down menu below listing the table in the database. Go to the `FacilityUser` table there.

When you need a Learner to "have no password" - then you will need to edit the `password` column for that Learner in the DB Browser. Single-click to highlight then type to edit the cell directly to the value `NOT_SPECIFIED` and click "Write Changes" at the top. Now that Learner "has no password".

*Things to Test Once*

- Turn off "Learners can see content without logging in" (aka turn off guest access). Then go to a URL pointing to a specific piece of content - which should take you to the Sign In page. Log in and make sure you are still redirected to the content.
- Check that you can complete the Learner who has no password scenario even when the Facility Setting "Learners can change their own password" is turned off.

*Things to Test In Each Scenario*

- Login as an Admin.
- Login as a Learner who has no password. Create a password & sign in. Logout. Log back in with the password you made.
- Login as a Learner who has a password.
- Change a Learner who has no password to a Coach or Admin & login as that user (create a password and finish the form).
- Change an Admin or Coach to a Learner and login as that Learner.

*Scenarios*

_You have <= 16 users of all kinds in your facility and are NOT in app context._

In addition to the above "every scenario" items:

- You should be presented first with a Username text field. Entering a username and clicking next will take you to password entry.
- Test invalid usernames and passwords.
- Test going back and forth between password and username fields.

_You have <= 16 users of all kinds in your facility and ARE in app context._

In addition to the above "every scenario" items:

- You should see the users listed as buttons.
- Click a user should work the same as the previous scenario as if you entered a username and act the same from there.
- Test going back and forth between password and users list fields

_You have > 16 users of all kinds in your facility and ARE in app context._

- You should see the same username text field flow as the first scenario.

### References


[Figma](https://www.figma.com/file/LI2L6gUncmCDDtaDECcbY6/Log-in-updates?node-id=147%3A5)

[Notion](https://www.notion.so/learningequality/Log-in-and-account-management-updates-0448fdd991c34c8e932c3626711d481b)

----

### Contributor Checklist


PR process:


Testing:


### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md
